In [1]:
import tensorflow as tf 
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 

Now that we're set with all the necessary imports, we go on to construct our model

In [9]:
input = tf.keras.Input(shape=(28, 28, 1), name = 'input_layer')
x = tf.keras.layers.Conv2D(filters = 15, 
                            kernel_size = 3, 
                            input_shape = (28, 28, 1), 
                            name = 'conv_layer_1') (input)
x = tf.keras.layers.MaxPooling2D(pool_size = 2, name = 'pooling_layer_1') (x)
x = tf.keras.layers.Conv2D(filters = 12, 
                            kernel_size = 3, 
                            input_shape = (13, 13, 10), 
                            name = 'conv_layer_2') (x)
x = tf.keras.layers.MaxPooling2D(pool_size = 2, name = 'pooling_layer_2') (x)
x = tf.keras.layers.Flatten(name = 'flatten_layer') (x)
x = tf.keras.layers.Dense(100, activation = 'relu', name = 'dense_hidden_layer') (x)
output = tf.keras.layers.Dense(10, activation = 'softmax', name = 'output_layer') (x)

model = tf.keras.Model(inputs = input, outputs = output, name = 'mnist_model')


print(model.summary())

Model: "mnist_model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (InputLayer)     [(None, 28, 28, 1)]       0         
_________________________________________________________________
conv_layer_1 (Conv2D)        (None, 26, 26, 15)        150       
_________________________________________________________________
pooling_layer_1 (MaxPooling2 (None, 13, 13, 15)        0         
_________________________________________________________________
conv_layer_2 (Conv2D)        (None, 11, 11, 12)        1632      
_________________________________________________________________
pooling_layer_2 (MaxPooling2 (None, 5, 5, 12)          0         
_________________________________________________________________
flatten_layer (Flatten)      (None, 300)               0         
_________________________________________________________________
dense_hidden_layer (Dense)   (None, 100)               

Our model consists of 2 convolutional layers, the first comprising of 15 kernels and the second one comprising of 12. In both these layers, the kernel size is the same, 3. We finally have a flattening layer and the output layer. We have added a dense layer to increase parameterisation, this has, in turn, increased the accuracy of the model.

In [10]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
x_train = tf.expand_dims(x_train, 3)
x_test = tf.expand_dims(x_test, 3)
x_test, x_val = tf.split(x_test, 2)
y_test, y_val = tf.split(y_test, 2)
model.compile(loss = 'sparse_categorical_crossentropy',
                optimizer = tf.keras.optimizers.Adam(),
                metrics = ['accuracy'])

We load the data set and expand the features by one dimension since the convolution layer takes in a 4-tensor while the features are 3-tensors. We use the *Adam optimizer* and *sparse categorical cross-entropy* as our loss function. We use *accuracy* as the metric of the model. 

In [12]:
history = model.fit(x_train, y_train, epochs = 5, validation_data = (x_val, y_val))
results = model.evaluate(x_test, y_test, verbose = 0)
print("Test Loss, Test Accuracy : ", results)

Train on 60000 samples, validate on 5000 samples
Epoch 1/5
60000/60000 [==============================] - 37s 617us/sample - loss: 0.0610 - accuracy: 0.9828 - val_loss: 0.0488 - val_accuracy: 0.9850
Epoch 2/5
60000/60000 [==============================] - 37s 613us/sample - loss: 0.0575 - accuracy: 0.9836 - val_loss: 0.0360 - val_accuracy: 0.9902
Epoch 3/5
60000/60000 [==============================] - 37s 611us/sample - loss: 0.0482 - accuracy: 0.9859 - val_loss: 0.0328 - val_accuracy: 0.9914
Epoch 4/5
60000/60000 [==============================] - 37s 624us/sample - loss: 0.0468 - accuracy: 0.9866 - val_loss: 0.0421 - val_accuracy: 0.9898
Epoch 5/5
60000/60000 [==============================] - 64s 1ms/sample - loss: 0.0470 - accuracy: 0.9870 - val_loss: 0.0378 - val_accuracy: 0.9898
Test Loss, Test Accuracy :  [0.12130467950571329, 0.9732]


We train the model for 5 epochs and achieve a test accuracy of 97.32%.